# Ćwiczenia z pandas

### Zadanie 1. Utwórz obiekt typu series na podstawie listy.

In [4]:
import pandas as pd
stocks = ['PLW', 'CDR', '11B', 'TEN']
print(pd.Series(stocks))

0    PLW
1    CDR
2    11B
3    TEN
dtype: object


### Zadanie 2. Utwórz obiekt typu series na podstawie słownika.

In [5]:
stocks = {'PLW': 387.00, 'CDR': 339.5, 'TEN': 349.5, '11B': 391.0}
print(pd.Series(stocks))

PLW    387.0
CDR    339.5
TEN    349.5
11B    391.0
dtype: float64


### Zadanie 3. Przekształć obiekt typu series na listę.

In [99]:
stocks = {'PLW': 387.00, 'CDR': 339.5, 'TEN': 349.5, '11B': 391.0}
quotations = pd.Series(data=stocks)
quotations = quotations.to_list()
quotations

[387.0, 339.5, 349.5, 391.0]

### Zadanie 4. Utwórz obiekt typu dataframe na podstawie obiektu typu series.

In [8]:
stocks = {'PLW': 387.00, 'CDR': 339.5, 'TEN': 349.5, '11B': 391.0}
quotations = pd.Series(data=stocks)
quotations = pd.DataFrame(quotations, columns=['price'])
print(quotations)

     price
PLW  387.0
CDR  339.5
TEN  349.5
11B  391.0


### Zadanie 5. Utwórz obiekt typu series z zadanymi indeksami.

In [12]:
import numpy as np
s = pd.Series(data=range(10, 100, 10), index=range(101, 110), dtype='float')
print(s)

101    10.0
102    20.0
103    30.0
104    40.0
105    50.0
106    60.0
107    70.0
108    80.0
109    90.0
dtype: float64


### Zadanie 6. Zmień typ danych na int.

In [15]:
series = pd.Series(['001', '002', '003', '004'], list('abcd'))
series = series.astype(int)
print(series)

a    1
b    2
c    3
d    4
dtype: int32


### Zadanie 7. Dodaj elementy do obiektu typu Series.

In [19]:
stocks = {'PLW': 387.00, 'CDR': 339.5, 'TEN': 349.5, '11B': 391.0}
quotations = pd.Series(data=stocks)
quotations = quotations.append(pd.Series({'BBT':25.5, 'F51':19.2}))
print(quotations)

PLW    387.0
CDR    339.5
TEN    349.5
11B    391.0
BBT     25.5
F51     19.2
dtype: float64


C:\Users\kubeq\AppData\Local\Temp\ipykernel_12064\3540210119.py:3: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  quotations = quotations.append(pd.Series({'BBT':25.5, 'F51':19.2}))


### Zadanie 8. Zresetuj indeks w obiekcie typu dataframe.

In [25]:
stocks = {'PLW': 387.00, 'CDR': 339.5, 'TEN': 349.5, '11B': 391.0, 'BBT': 25.5, 'F51': 19.2}
quotations = pd.Series(data=stocks)
quotations = pd.DataFrame(quotations).reset_index()
quotations.columns = ['ticker', 'price']
print(quotations)

  ticker  price
0    PLW  387.0
1    CDR  339.5
2    TEN  349.5
3    11B  391.0
4    BBT   25.5
5    F51   19.2


### Zadanie 9. Utwórz obiekt typu DataFrame.

In [34]:
companies = pd.DataFrame({'company': ['Amazon', 'Microsoft', 'Facebook'], 
                         'price': [2375.0, 178.6, 179.2], 
                         'ticker': ['AMZN.US','MSFT.US','FB.US']})
companies

company   price   ticker
0     Amazon  2375.0  AMZN.US
1  Microsoft   178.6  MSFT.US
2   Facebook   179.2    FB.US

### Zadanie 10. Przekształć kolumnę company na indeks.

In [35]:
data_dict = {
    'company': ['Amazon', 'Microsoft', 'Facebook'],
    'price': [2375.00, 178.6, 179.2],
    'ticker': ['AMZN.US', 'MSFT.US', 'FB.US']
}
companies = pd.DataFrame(data=data_dict)
companies = companies.set_index('company')
print(companies)

            price   ticker
company                   
Amazon     2375.0  AMZN.US
Microsoft   178.6  MSFT.US
Facebook    179.2    FB.US


### Zadanie 11. Zbuduj obiekt zawierający daty dla wszystkich dni od stycznia 2020.

In [36]:
date_range = pd.date_range(start='2020-01-01', periods=31)
print(date_range)

DatetimeIndex(['2020-01-01', '2020-01-02', '2020-01-03', '2020-01-04',
               '2020-01-05', '2020-01-06', '2020-01-07', '2020-01-08',
               '2020-01-09', '2020-01-10', '2020-01-11', '2020-01-12',
               '2020-01-13', '2020-01-14', '2020-01-15', '2020-01-16',
               '2020-01-17', '2020-01-18', '2020-01-19', '2020-01-20',
               '2020-01-21', '2020-01-22', '2020-01-23', '2020-01-24',
               '2020-01-25', '2020-01-26', '2020-01-27', '2020-01-28',
               '2020-01-29', '2020-01-30', '2020-01-31'],
              dtype='datetime64[ns]', freq='D')


### Zadanie 12. Zbuduj obiekt zawierający daty dla wszystkich poniedziałków od stycznia 2020.

In [ ]:
import pandas as pd
date_range = pd.date_range(start='2020-01-01', end='2020-12-31', freq='W-MON')
print(date_range)

### Zadanie 13. Zbuduj obiekt zawierający godziny określonego dnia.

In [38]:
date_range = pd.date_range(start='2021-01-01', periods=24, freq='H')
print(date_range)

DatetimeIndex(['2021-01-01 00:00:00', '2021-01-01 01:00:00',
               '2021-01-01 02:00:00', '2021-01-01 03:00:00',
               '2021-01-01 04:00:00', '2021-01-01 05:00:00',
               '2021-01-01 06:00:00', '2021-01-01 07:00:00',
               '2021-01-01 08:00:00', '2021-01-01 09:00:00',
               '2021-01-01 10:00:00', '2021-01-01 11:00:00',
               '2021-01-01 12:00:00', '2021-01-01 13:00:00',
               '2021-01-01 14:00:00', '2021-01-01 15:00:00',
               '2021-01-01 16:00:00', '2021-01-01 17:00:00',
               '2021-01-01 18:00:00', '2021-01-01 19:00:00',
               '2021-01-01 20:00:00', '2021-01-01 21:00:00',
               '2021-01-01 22:00:00', '2021-01-01 23:00:00'],
              dtype='datetime64[ns]', freq='H')


### Zadanie 14. Zbuduj obiekt DataFrame z datami i numerami dnia w ciągu roku.

In [61]:
day = pd.date_range(start = '2021-03-01', end = '2021-03-31')
df = pd.DataFrame({'day':day})
df['day_of_year'] = df['day'].dt.dayofyear
print(df[:5])

         day  day_of_year
0 2021-03-01           60
1 2021-03-02           61
2 2021-03-03           62
3 2021-03-04           63
4 2021-03-05           64


### Zadanie 15. Zbuduj obiekt DataFrame jako indeks ustawiając daty.

In [43]:
np.random.seed(42)
data_dict = {
    'normal': np.random.normal(loc=0, scale=1, size=1000),
    'uniform': np.random.uniform(low=0, high=1, size=1000),
    'binomial': np.random.binomial(n=1, p=0.2, size=1000)
}
indeks = pd.date_range(start = '2020-01-01', periods=1000)
df = pd.DataFrame(data_dict, indeks)
print(df)

              normal   uniform  binomial
2020-01-01  0.496714  0.167483         0
2020-01-02 -0.138264  0.104568         0
2020-01-03  0.647689  0.636430         0
2020-01-04  1.523030  0.706476         0
2020-01-05 -0.234153  0.031586         1
...              ...       ...       ...
2022-09-22 -0.281100  0.280683         1
2022-09-23  1.797687  0.206993         0
2022-09-24  0.640843  0.516573         0
2022-09-25 -0.571179  0.005506         1
2022-09-26  0.572583  0.007664         0

[1000 rows x 3 columns]


### Zadanie 16. Wydrukuj pierwsze 10 i ostatnie 5 wierszy obiektu typu DataFrame.

In [49]:
print(df[:10])
print()
print(df[-5:])

              normal   uniform  binomial
2020-01-01  0.496714  0.167483         0
2020-01-02 -0.138264  0.104568         0
2020-01-03  0.647689  0.636430         0
2020-01-04  1.523030  0.706476         0
2020-01-05 -0.234153  0.031586         1
2020-01-06 -0.234137  0.936212         0
2020-01-07  1.579213  0.051971         0
2020-01-08  0.767435  0.541296         0
2020-01-09 -0.469474  0.709061         1
2020-01-10  0.542560  0.870969         0

              normal   uniform  binomial
2022-09-22 -0.281100  0.280683         1
2022-09-23  1.797687  0.206993         0
2022-09-24  0.640843  0.516573         0
2022-09-25 -0.571179  0.005506         1
2022-09-26  0.572583  0.007664         0


### Zadanie 17. Wydrukuj statystyki opisowe zmiennych z obiektu typu DataFrame.

In [50]:
print(df.info())
print()
print(df.describe())

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1000 entries, 2020-01-01 to 2022-09-26
Freq: D
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   normal    1000 non-null   float64
 1   uniform   1000 non-null   float64
 2   binomial  1000 non-null   int32  
dtypes: float64(2), int32(1)
memory usage: 27.3 KB
None

            normal      uniform    binomial
count  1000.000000  1000.000000  1000.00000
mean      0.019332     0.503649     0.19200
std       0.979216     0.288357     0.39407
min      -3.241267     0.003218     0.00000
25%      -0.647590     0.247199     0.00000
50%       0.025301     0.516126     0.00000
75%       0.647944     0.746319     0.00000
max       3.852731     0.999414     1.00000


### Zadanie 18. Znajdź rozkład zmiennej z obiektu typu DataFrame.

In [51]:
print(df['binomial'].value_counts())

0    808
1    192
Name: binomial, dtype: int64


### Zadanie 19. Zapisz 50 pierwszych wierszy do pliku csv.

In [56]:
df[:50].to_csv('dataframe50.csv')

### Zadanie 20. Wczytaj zapisany plik pamiętaj o indeksie.

In [4]:
df_new = pd.read_csv('dataframe50.csv', index_col=0)
df_new[:5]

normal   uniform  binomial
2020-01-01  0.496714  0.167483         0
2020-01-02 -0.138264  0.104568         0
2020-01-03  0.647689  0.636430         0
2020-01-04  1.523030  0.706476         0
2020-01-05 -0.234153  0.031586         1

### Zadanie 21. Wczytaj plik.

In [ ]:
google = pd.read_csv("google.csv")
print(google)

### Zadanie 22. Wypisz statystyki opisowe zmiennych z wczytanego pliku.

In [ ]:
print(google.info())
print()
print(google.describe())

### Zadanie 23. Zamień indeks na kolumnę.

In [ ]:
google = google.reset_index()
print(google)

### Zadanie 24. Dodaj kolumnę z miesiącem i rokiem.

In [ ]:
google['Date'] = pd.to_datetime(google['Date'])
google['Year'] = google['Date'].dt.year
google['Month'] = google['Date'].dt.month

### Zadanie 25. Znajdź średnią cenę zamknięcia dla poszczególnych miesięcy.

In [ ]:
print(google.groupby('Month')['Close'].mean())

### Zadanie 26. Znajdź wiersz z minimalną ceną zmaknięcia.

In [ ]:
idx_min = google['Close'].argmin()
print(google.iloc[[idx_min]])

### Zadanie 27. Wyodrębnij określone kolumny z obiektu typu DataFrame.

In [ ]:
print(google[['Date','Open','Close','Volume']])

### Zadanie 28. Ustaw określony indeks kolumny.

In [ ]:
google = google.set_index('Date')

### Zadanie 29. Usuń określone kolumny z obiketu typu DataFrame.

In [ ]:
google = google.drop(columns=['Year', 'Month'])

### Zadanie 30. Zmień nazwy kolumn.

In [ ]:
google.columns = ['O', 'H', 'L', 'C', 'V']

### Zadanie 31. Wczytaj plik i wyświetl 5 pierwszych wierszy.

In [64]:
url = 'https://ml-repository-krakers.s3-eu-west-1.amazonaws.com/kaggle+/churn_modelling/Telco-Customer-Churn.csv'
df = pd.read_csv(url)
df[:5]

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

### Zadanie 32. Wyświetl liczbę brakujacych danych z każdej z kolumn.

In [65]:
print(df.isnull().sum())

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


### Zadanie 33. Usuń spacje, zastąp je wartością mediany.

In [68]:
TotalChargesMedian = df['TotalCharges'][df['TotalCharges'] != ' '].median()
df.loc[df['TotalCharges'] == ' ', 'TotalCharges'] = TotalChargesMedian
df['TotalCharges'] = df['TotalCharges'].astype('float')
print(df['TotalCharges'].value_counts())

1397.475    11
20.200      11
19.750       9
20.050       8
19.900       8
            ..
6849.400     1
692.350      1
130.150      1
3211.900     1
6844.500     1
Name: TotalCharges, Length: 6531, dtype: int64


### Zadanie 34. Przypisz zmiennym określone typy.

In [71]:
categorical = ['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 
               'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
               'StreamingTV', 'Contract', 'StreamingMovies', 'PaperlessBilling', 'PaymentMethod', 'Churn'] 
 
numerical = ['tenure', 'MonthlyCharges']
 
for col in categorical:
    df[col] = pd.Categorical(df[col]) 
for col in numerical:
    df[col] = df[col].astype(float)
df.info()  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   customerID        7043 non-null   object  
 1   gender            7043 non-null   category
 2   SeniorCitizen     7043 non-null   category
 3   Partner           7043 non-null   category
 4   Dependents        7043 non-null   category
 5   tenure            7043 non-null   float64 
 6   PhoneService      7043 non-null   category
 7   MultipleLines     7043 non-null   category
 8   InternetService   7043 non-null   category
 9   OnlineSecurity    7043 non-null   category
 10  OnlineBackup      7043 non-null   category
 11  DeviceProtection  7043 non-null   category
 12  TechSupport       7043 non-null   category
 13  StreamingTV       7043 non-null   category
 14  StreamingMovies   7043 non-null   category
 15  Contract          7043 non-null   category
 16  PaperlessBilling  7043 n

### Zadanie 35. Wyświetl charakterystyki opisowe zmiennych kateogrycznych.

In [76]:
df.describe(include=['category'])

gender  SeniorCitizen Partner Dependents PhoneService MultipleLines  \
count    7043           7043    7043       7043         7043          7043   
unique      2              2       2          2            2             3   
top      Male              0      No         No          Yes            No   
freq     3555           5901    3641       4933         6361          3390   

       InternetService OnlineSecurity OnlineBackup DeviceProtection  \
count             7043           7043         7043             7043   
unique               3              3            3                3   
top        Fiber optic             No           No               No   
freq              3096           3498         3088             3095   

       TechSupport StreamingTV StreamingMovies        Contract  \
count         7043        7043            7043            7043   
unique           3           3               3               3   
top             No          No              No  Month-to-month   
freq          3473        2810            2785            3875   

       PaperlessBilling     PaymentMethod Churn  
count              7043              7043  7043  
unique                2                 4     2  
top                 Yes  Electronic check    No  
freq               4171              2365  5174

### Zadanie 36. Wyświetl rozkład wybranej zmiennej.

In [77]:
print(df['Churn'].value_counts())

No     5174
Yes    1869
Name: Churn, dtype: int64


### Zadanie 37. Wyświetl średnie dla podgrup.

In [82]:
print(df.groupby(['Churn','PaymentMethod'])['MonthlyCharges'].mean())

Churn  PaymentMethod            
No     Bank transfer (automatic)    65.049417
       Credit card (automatic)      64.562209
       Electronic check             74.232032
       Mailed check                 41.403911
Yes    Bank transfer (automatic)    77.875581
       Credit card (automatic)      77.356034
       Electronic check             78.700980
       Mailed check                 54.557143
Name: MonthlyCharges, dtype: float64


### Zadanie 38. Zmień Yes na 1 oraz No na 0

In [ ]:
df['Churn'] = df['Churn'].map({'Yes': 1, 'No': 0})

### Zadanie 39. Wyświetl macierz korelacji.

In [83]:
print(df.corr())

                  tenure  MonthlyCharges  TotalCharges
tenure          1.000000        0.247900      0.825464
MonthlyCharges  0.247900        1.000000      0.650864
TotalCharges    0.825464        0.650864      1.000000


### Zadanie 40. Pobierz próbkę 10 wierszy i zapisz do pliku csv.

In [ ]:
df.sample(10).to_csv('sample_10.csv')

### Zadanie 41. Z dwóch obiektów typu series utwórz obiekt typu DataFrame.

In [84]:
np.random.seed(42)
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))
df = pd.DataFrame({'col1':s1, 'col2':s2})
print(df)

        col1      col2
0   0.374540 -1.012831
1   0.950714  0.314247
2   0.731994 -0.908024
3   0.598658 -1.412304
4   0.156019  1.465649
5   0.155995 -0.225776
6   0.058084  0.067528
7   0.866176 -1.424748
8   0.601115 -0.544383
9   0.708073  0.110923
10  0.020584 -1.150994
11  0.969910  0.375698
12  0.832443 -0.600639
13  0.212339 -0.291694
14  0.181825 -0.601707
15  0.183405  1.852278
16  0.304242 -0.013497
17  0.524756 -1.057711
18  0.431945  0.822545
19  0.291229 -1.220844


### Zadanie 42. Wyodrębnij określone wiersze z obiektu typu DataFrame.

In [86]:
print(df[(df['col2'] >= 0.0) & (df['col2'] <= 1.0)])

        col1      col2
1   0.950714  0.314247
6   0.058084  0.067528
9   0.708073  0.110923
11  0.969910  0.375698
18  0.431945  0.822545


### Zadanie 43. Stwórz nową kolumnę.

In [87]:
df['col3'] = df['col2'].map(lambda x: 1 if x >= 0 else -1)
print(df)

        col1      col2  col3
0   0.374540 -1.012831    -1
1   0.950714  0.314247     1
2   0.731994 -0.908024    -1
3   0.598658 -1.412304    -1
4   0.156019  1.465649     1
5   0.155995 -0.225776    -1
6   0.058084  0.067528     1
7   0.866176 -1.424748    -1
8   0.601115 -0.544383    -1
9   0.708073  0.110923     1
10  0.020584 -1.150994    -1
11  0.969910  0.375698     1
12  0.832443 -0.600639    -1
13  0.212339 -0.291694    -1
14  0.181825 -0.601707    -1
15  0.183405  1.852278     1
16  0.304242 -0.013497    -1
17  0.524756 -1.057711    -1
18  0.431945  0.822545     1
19  0.291229 -1.220844    -1


### Zadanie 44. Stwórz nową kolumnę (obetnij wartości z góry i z dołu).

In [89]:
df['col4'] = df['col2'].map(lambda x: 1 if x>1 else x).map(lambda x: -1 if x<-1 else x)

### Zadanie 45. Wyświetl najwyższe i najniższe wartości w danej kolumnie.

In [90]:
print(df['col2'].nlargest(5))
print()
print(df['col2'].nsmallest(5))

15    1.852278
4     1.465649
18    0.822545
11    0.375698
1     0.314247
Name: col2, dtype: float64

7    -1.424748
3    -1.412304
19   -1.220844
10   -1.150994
17   -1.057711
Name: col2, dtype: float64


### Zadanie 46. Wyświetl skumulowane sumy dla każdej z kolumn.

In [91]:
print(df.cumsum())

        col1      col2  col3      col4
0   0.374540 -1.012831    -1 -1.000000
1   1.325254 -0.698584     0 -0.685753
2   2.057248 -1.606608    -1 -1.593777
3   2.655907 -3.018912    -2 -2.593777
4   2.811925 -1.553263    -1 -1.593777
5   2.967920 -1.779039    -2 -1.819553
6   3.026004 -1.711511    -1 -1.752025
7   3.892180 -3.136259    -2 -2.752025
8   4.493295 -3.680642    -3 -3.296408
9   5.201367 -3.569719    -2 -3.185485
10  5.221952 -4.720713    -3 -4.185485
11  6.191862 -4.345015    -2 -3.809787
12  7.024304 -4.945653    -3 -4.410426
13  7.236643 -5.237347    -4 -4.702119
14  7.418468 -5.839054    -5 -5.303826
15  7.601873 -3.986776    -4 -4.303826
16  7.906115 -4.000273    -5 -4.317323
17  8.430872 -5.057984    -6 -5.317323
18  8.862817 -4.235439    -5 -4.494778
19  9.154046 -5.456283    -6 -5.494778


### Zadanie 47. Wyświetl medianę dla wartości w danej kolumnie.

In [ ]:
print(df['col2'].median())

### Zadanie 48. Wyświetl określone wiersze.

In [92]:
print(df[df['col2'] > 0.0])

        col1      col2  col3      col4
1   0.950714  0.314247     1  0.314247
4   0.156019  1.465649     1  1.000000
6   0.058084  0.067528     1  0.067528
9   0.708073  0.110923     1  0.110923
11  0.969910  0.375698     1  0.375698
15  0.183405  1.852278     1  1.000000
18  0.431945  0.822545     1  0.822545


### Zadanie 49. Przekotwertuj obiekt typu DataFrame na słownik.

In [96]:
df[:5].to_dict()

{'col1': {0: 0.3745401188473625,
  1: 0.9507143064099162,
  2: 0.7319939418114051,
  3: 0.5986584841970366,
  4: 0.15601864044243652},
 'col2': {0: -1.0128311203344238,
  1: 0.3142473325952739,
  2: -0.9080240755212109,
  3: -1.4123037013352915,
  4: 1.465648768921554},
 'col3': {0: -1, 1: 1, 2: -1, 3: -1, 4: 1},
 'col4': {0: -1.0,
  1: 0.3142473325952739,
  2: -0.9080240755212109,
  3: -1.0,
  4: 1.0}}

### Zadanie 50. Przekotwertuj obiekt typu DataFrame na obiekt typu html.

In [98]:
x = df[:5].to_html()

### Zadanie 51. Wyodrębnij określone wiersze.

In [102]:
np.random.seed(42)
df = pd.DataFrame(np.random.rand(10, 4), columns=list('ABCD'))
print(df[df['C']>0.8])

          A         B         C         D
8  0.065052  0.948886  0.965632  0.808397


### Zadanie 52. Wyodrębnij określone wiersze (2 warunki).

In [105]:
print(df[(df['C']>0.3) & (df['D']<0.7)])

          A         B         C         D
0  0.374540  0.950714  0.731994  0.598658
4  0.304242  0.524756  0.431945  0.291229
7  0.592415  0.046450  0.607545  0.170524
9  0.304614  0.097672  0.684233  0.440152


### Zadanie 53. Iteracyjnie wypisz 5 pierwszych wierszy.

In [106]:
for index, row in df[:5].iterrows():
    print(row)

A    0.374540
B    0.950714
C    0.731994
D    0.598658
Name: 0, dtype: float64
A    0.156019
B    0.155995
C    0.058084
D    0.866176
Name: 1, dtype: float64
A    0.601115
B    0.708073
C    0.020584
D    0.969910
Name: 2, dtype: float64
A    0.832443
B    0.212339
C    0.181825
D    0.183405
Name: 3, dtype: float64
A    0.304242
B    0.524756
C    0.431945
D    0.291229
Name: 4, dtype: float64


### Zadanie 54. Zmień wartości określonych pól.

In [114]:
df.loc[3, 'B'] = np.nan
df.loc[8, 'D'] = np.nan

### Zadanie 55. Usuń wiersze z brakami danych.

In [115]:
df1 = df.dropna()

### Zadanie 56. Zresetuj indeks.

In [120]:
df1 = df1.reset_index(drop = True)

In [119]:
df1

index         A         B         C         D
0      0  0.374540  0.950714  0.731994  0.598658
1      1  0.156019  0.155995  0.058084  0.866176
2      2  0.601115  0.708073  0.020584  0.969910
3      3  0.304242  0.524756  0.431945  0.291229
4      4  0.611853  0.139494  0.292145  0.366362
5      5  0.456070  0.785176  0.199674  0.514234
6      6  0.592415  0.046450  0.607545  0.170524
7      7  0.304614  0.097672  0.684233  0.440152

### Zadanie 57. Określ liczbę brakujących elementów w kolumnach.

In [121]:
print(df.isnull().sum())

A    0
B    1
C    0
D    1
dtype: int64


### Zadanie 58. Wypełnij zerami brakujące pola.

In [122]:
df = df.fillna(0)

### Zadanie 59. Zamień kolejność kolumn.

In [123]:
df = df[['D', 'A', 'B', 'C']]
print(df)

          D         A         B         C
0  0.598658  0.374540  0.950714  0.731994
1  0.866176  0.156019  0.155995  0.058084
2  0.969910  0.601115  0.708073  0.020584
3  0.183405  0.832443  0.000000  0.181825
4  0.291229  0.304242  0.524756  0.431945
5  0.366362  0.611853  0.139494  0.292145
6  0.514234  0.456070  0.785176  0.199674
7  0.170524  0.592415  0.046450  0.607545
8  0.000000  0.065052  0.948886  0.965632
9  0.440152  0.304614  0.097672  0.684233


### Zadanie 60. Usuń jedną z kolumn.

In [128]:
df = df.drop(columns = 'D')
print(df)

          A         B         C
0  0.374540  0.950714  0.731994
1  0.156019  0.155995  0.058084
2  0.601115  0.708073  0.020584
3  0.832443  0.000000  0.181825
4  0.304242  0.524756  0.431945
5  0.611853  0.139494  0.292145
6  0.456070  0.785176  0.199674
7  0.592415  0.046450  0.607545
8  0.065052  0.948886  0.965632
9  0.304614  0.097672  0.684233


### Zadanie 61. Wczytaj dane, ustaw indeks.

In [53]:
import pandas as pd
url = 'https://storage.googleapis.com/esmartdata-courses-files/dash-course/data.csv'
df = pd.read_csv(url, index_col=0)

### Zadanie 62. Wczytaj dane, ustaw indeks.

In [54]:
print(list(df.columns))

['Name', 'Location', 'Year', 'Kilometers_Driven', 'Fuel_Type', 'Transmission', 'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats', 'New_Price', 'Price']


### Zadanie 63. Usuń wybraną kolumnę.

In [55]:
df=df.drop(columns="New_Price")
print(df[:5])

                               Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   Price  
0   1.75  
1  12.50  
2   4.50  
3   6.00  
4  17.74  


### Zadanie 64. Znajdź liczbę brakujących informacji.

In [56]:
df.isnull().sum()

Name                  0
Location              0
Year                  0
Kilometers_Driven     0
Fuel_Type             0
Transmission          0
Owner_Type            0
Mileage               2
Engine               36
Power                36
Seats                42
Price                 0
dtype: int64

### Zadanie 65. Usuń wiersze z brakami danych.

In [57]:
df.dropna(inplace = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5975 entries, 0 to 6018
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               5975 non-null   object 
 1   Location           5975 non-null   object 
 2   Year               5975 non-null   int64  
 3   Kilometers_Driven  5975 non-null   int64  
 4   Fuel_Type          5975 non-null   object 
 5   Transmission       5975 non-null   object 
 6   Owner_Type         5975 non-null   object 
 7   Mileage            5975 non-null   object 
 8   Engine             5975 non-null   object 
 9   Power              5975 non-null   object 
 10  Seats              5975 non-null   float64
 11  Price              5975 non-null   float64
dtypes: float64(2), int64(2), object(8)
memory usage: 606.8+ KB


### Zadanie 66. Zastąp duże litery nazw kolumn małymi.

In [58]:
df.columns = [col.lower() for col in df.columns]

### Zadanie 67. Wyświetl rozkład wybranej zmiennej.

In [59]:
df['owner_type'].value_counts()

First             4903
Second             953
Third              111
Fourth & Above       8
Name: owner_type, dtype: int64

### Zadanie 68. Usuń 3 ostatnie znaki z pól wybranej kolumny.

In [60]:
df['engine'] = df['engine'].map(lambda x: int(x[:-3]))
print(df[['name','engine']][:5])

                               name  engine
0            Maruti Wagon R LXI CNG     998
1  Hyundai Creta 1.6 CRDi SX Option    1582
2                      Honda Jazz V    1199
3                 Maruti Ertiga VDI    1248
4   Audi A4 New 2.0 TDI Multitronic    1968


### Zadanie 69. Wydrukuj 5 najczęstszych wartości okreslonej zmiennej kategorycznej.

In [62]:
df['power'] = df['power'].map(lambda x: np.nan if x=='null nhp' else x)
print(df['power'].value_counts()[:5])

74 bhp      235
98.6 bhp    131
73.9 bhp    125
140 bhp     123
78.9 bhp    111
Name: power, dtype: int64

### Zadanie 70. Sprawdź ile jest obserwacji dla poszczególnych lat.

In [67]:
df.groupby('year').size()[:5]

year
1998     4
1999     2
2000     4
2001     7
2002    14
dtype: int64

### Zadanie 71. Dokonaj mapowania zmiennej.

In [70]:
url = 'https://storage.googleapis.com/esmartdata-courses-files/dash-course/data.csv'
df = pd.read_csv(url, index_col=0)
df.columns = [col.lower() for col in df.columns]
df['transmission'].value_counts()
df['transmission'] = df['transmission'].map(lambda x: 0 if x=='Manual' else  1)
print(df['transmission'][:5])

0    0
1    0
2    0
3    0
4    1
Name: transmission, dtype: int64


### Zadanie 72. Zapisz obiekt typu DataFrame do pliku csv. Nie zapisuj indeksu.

In [71]:
df.to_csv("cars.csv", index=False)

### Zadanie 73. Wczytaj obiket typu DataFrame.

In [72]:
url = 'https://storage.googleapis.com/esmartdata-courses-files/ds-bootcamp/london_bike.csv'
df = pd.read_csv(url, index_col=0)
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17414 entries, 0 to 17413
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   timestamp     17414 non-null  object 
 1   cnt           17414 non-null  int64  
 2   t1            17414 non-null  float64
 3   t2            17414 non-null  float64
 4   hum           17414 non-null  float64
 5   wind_speed    17414 non-null  float64
 6   weather_code  17414 non-null  float64
 7   is_holiday    17414 non-null  float64
 8   is_weekend    17414 non-null  float64
 9   season        17414 non-null  float64
dtypes: float64(8), int64(1), object(1)
memory usage: 1.3+ MB
None


### Zadanie 74. Wyświetl 5 pierwszych i 5 ostatnich wierszy obiektu typu DataFrame.

In [ ]:
print(df[:5])
print(df[-5:])

### Zadanie 75. Sformatuj datę.

In [74]:
df["timestamp"] = pd.to_datetime(df["timestamp"])
df.info()

### Zadanie 76. Utwórz kolumnę z godziną.

In [77]:
df['hour']=df["timestamp"].dt.hour

### Zadanie 77. Utwórz kolumnę z miesiącem..

In [80]:
df['month'] = df['timestamp'].dt.month

### Zadanie 78. Oblicz średnią dla miesięcy.

In [84]:
humidity_by_month = df.groupby('month')['hum'].mean().reset_index()
print(humidity_by_month)

    month        hum
0       1  77.543376
1       2  74.905813
2       3  71.903781
3       4  67.532337
4       5  66.367272
5       6  68.945148
6       7  64.383525
7       8  67.289420
8       9  71.753228
9      10  77.639621
10     11  78.732517
11     12  81.032345


### Zadanie 79. Oblicz średnią dla godzin.

In [87]:
df.groupby('hour')['cnt'].mean().reset_index()[:4]

hour         cnt
0     0  290.609116
1     1  200.631215
2     2  136.303745
3     3   94.245492

### Zadanie 80. Oblicz średnią dla podgrup.

In [93]:
df.groupby(['is_weekend','hour'])['cnt'].mean().reset_index()[:5]

is_weekend  hour         cnt
0         0.0     0  215.366795
1         0.0     1  128.001931
2         0.0     2   81.007737
3         0.0     3   53.250000
4         0.0     4   52.248062

### Zadanie 81. Wyodrębnij określone wiersze (2 warunku)

In [95]:
print(df[(df['wind_speed']<10.0) & (df['hum']>90.0)][:1])

   timestamp  cnt   t1   t2   hum  wind_speed  weather_code  is_holiday  \
0 2015-01-04  182  3.0  2.0  93.0         6.0           3.0         0.0   

   is_weekend  season  hour  month  
0         1.0     3.0     0      1  


### Zadanie 82. Wyodrębnij określone wiersze i przygotuj nowy obiekt typu DataFrame.

In [97]:
df_weekend = df[df['is_weekend']==1.0]

### Zadanie 83. Zapisz obiekt typu DataFrame do pliku csv.

In [107]:
df_weekend.to_csv('weekend.txt', index=False)

### Zadanie 84. Wczytaj z pliku csv obiekt typu DataFrame.

In [108]:
url = 'https://storage.googleapis.com/esmartdata-courses-files/ml-course/insurance.csv'
df = pd.read_csv(url)

### Zadanie 85. Sprawdź czy w obiekcie typu DataFrame są dupikaty.

In [109]:
url = 'https://storage.googleapis.com/esmartdata-courses-files/ml-course/insurance.csv'
df = pd.read_csv(url)
print(df[df.duplicated()])

     age   sex    bmi  children smoker     region    charges
581   19  male  30.59         0     no  northwest  1639.5631


### Zadanie 86. Usuń duplikaty.

In [106]:
df = df.drop_duplicates()

### Zadanie 87. Wyświetl kolumny typu object.

In [113]:
print(df.select_dtypes(include=['object']))

         sex smoker     region
0     female    yes  southwest
1       male     no  southeast
2       male     no  southeast
3       male     no  northwest
4       male     no  northwest
...      ...    ...        ...
1333    male     no  northwest
1334  female     no  northeast
1335  female     no  southeast
1336  female     no  southwest
1337  female    yes  northwest

[1338 rows x 3 columns]


### Zadanie 88. Zmień typ zmiennych.

In [115]:
categoriacal = ['sex', 'smoker', 'region']
for col in categoriacal:
    df[col] = df[col].astype('category')
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   age       1338 non-null   int64   
 1   sex       1338 non-null   category
 2   bmi       1338 non-null   float64 
 3   children  1338 non-null   int64   
 4   smoker    1338 non-null   category
 5   region    1338 non-null   category
 6   charges   1338 non-null   float64 
dtypes: category(3), float64(2), int64(2)
memory usage: 46.3 KB
None


### Zadanie 89. Wskaż zmienne numeryczne

In [118]:
num_vars = list(df.select_dtypes(include=['float', 'int']).columns)
print(num_vars)

['age', 'bmi', 'children', 'charges']


### Zadanie 90. Utwórz obiekt typu DataFrame wyłącznie ze zmiennymi numerycznymi / kateogrycznymi.

In [ ]:
df_cat = df.select_dtypes(include=['category'])
df_num = df.select_dtypes(include=['float','int'])

### Zadanie 91. Wyświetl statystyki opisowe.

In [2]:
import pandas as pd
url = 'https://storage.googleapis.com/esmartdata-courses-files/ml-course/insurance.csv'
df = pd.read_csv(url)
for col in list(df.select_dtypes(include=['object']).columns):
    df[col] = df[col].astype('category')
df_cat = df.select_dtypes(include=['category']).copy()
print(df_cat.describe())

         sex smoker     region
count   1338   1338       1338
unique     2      2          4
top     male     no  southeast
freq     676   1064        364


### Zadanie 92. Wyświetl statystyki opisowe.

In [ ]:
url = 'https://storage.googleapis.com/esmartdata-courses-files/ml-course/insurance.csv'
df = pd.read_csv(url)
for col in list(df.select_dtypes(include=['object']).columns):
    df[col] = df[col].astype('category')
df_num = df.select_dtypes(include=['float', 'int']).copy()

print(df_num.describe())

### Zadanie 93. Wyświetl wybrane statystyki opisowe.

In [3]:
df = pd.read_csv(url)
print(df.describe().T[['mean', 'std']])

                  mean           std
age          39.207025     14.049960
bmi          30.663397      6.098187
children      1.094918      1.205493
charges   13270.422265  12110.011237


### Zadanie 94. Sprawdź czy występują braki danych.

In [4]:
url = 'https://storage.googleapis.com/esmartdata-courses-files/ml-course/insurance.csv'
df = pd.read_csv(url)
df.isnull().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

### Zadanie 95. Przygotuj zmienne zerojedynkowe.

In [6]:
pd.set_option('display.width', 150)
url = 'https://storage.googleapis.com/esmartdata-courses-files/ml-course/insurance.csv'
df = pd.read_csv(url)
for col in list(df.select_dtypes(include=['object']).columns):
    df[col] = df[col].astype('category')
df_dummies = pd.get_dummies(df, drop_first=True)
print(df_dummies.head())

   age     bmi  children      charges  sex_male  smoker_yes  region_northwest  region_southeast  region_southwest
0   19  27.900         0  16884.92400         0           1                 0                 0                 1
1   18  33.770         1   1725.55230         1           0                 0                 1                 0
2   28  33.000         3   4449.46200         1           0                 0                 1                 0
3   33  22.705         0  21984.47061         1           0                 1                 0                 0
4   32  28.880         0   3866.85520         1           0                 1                 0                 0


### Zadanie 96. Oblicz korelacje między parami zmiennych.

In [8]:
corr = df_dummies.corr()

### Zadanie 97. Posortuj wiersze.

In [9]:
print(corr[['charges']].sort_values(by='charges', ascending=False))

                   charges
charges           1.000000
smoker_yes        0.787251
age               0.299008
bmi               0.198341
region_southeast  0.073982
children          0.067998
sex_male          0.057292
region_northwest -0.039905
region_southwest -0.043210


### Zadanie 98. Usuń wybraną kolumnę.

In [13]:
pd.set_option('display.width', 150)
url = 'https://storage.googleapis.com/esmartdata-courses-files/ml-course/insurance.csv'
df = pd.read_csv(url)
for col in list(df.select_dtypes(include=['object']).columns):
    df[col] = df[col].astype('category')
df_dummies = pd.get_dummies(df, drop_first=True)
data = df_dummies.copy()
target = data.pop("charges")
print(data[:5])
print()
print(df_dummies[:5])

   age     bmi  children  sex_male  smoker_yes  region_northwest  region_southeast  region_southwest
0   19  27.900         0         0           1                 0                 0                 1
1   18  33.770         1         1           0                 0                 1                 0
2   28  33.000         3         1           0                 0                 1                 0
3   33  22.705         0         1           0                 1                 0                 0
4   32  28.880         0         1           0                 1                 0                 0

   age     bmi  children      charges  sex_male  smoker_yes  region_northwest  region_southeast  region_southwest
0   19  27.900         0  16884.92400         0           1                 0                 0                 1
1   18  33.770         1   1725.55230         1           0                 0                 1                 0
2   28  33.000         3   4449.46200         1    

### Zadanie 99. Utwórz obiekt typu DataFrame na podstawie słownika.

In [31]:
data = {
    'size': ['XL', 'L', 'M', 'L', 'M'],
    'color': ['red', 'green', 'blue', 'green', 'red'],
    'gender': ['female', 'male', 'male', 'female', 'female'],
    'price': [199.0, 89.0, 99.0, 129.0, 79.0],
    'weight': [500, 450, 300, 380, 410],
    'bought': ['yes', 'no', 'yes', 'no', 'yes']
}
df = pd.DataFrame(data)
print(df)

  size  color  gender  price  weight bought
0   XL    red  female  199.0     500    yes
1    L  green    male   89.0     450     no
2    M   blue    male   99.0     300    yes
3    L  green  female  129.0     380     no
4    M    red  female   79.0     410    yes


### Zadanie 100. Zmień typy zmiennych.


In [32]:
print(df.info())
df['size'] = df['size'].astype('category')
df['color']=df['color'].astype('category')
df['gender'] = df['gender'].astype('category')
df['weight'] = df['weight'].astype('float')
df['bought'] = df['bought'].astype('category')
print()
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   size    5 non-null      object 
 1   color   5 non-null      object 
 2   gender  5 non-null      object 
 3   price   5 non-null      float64
 4   weight  5 non-null      int64  
 5   bought  5 non-null      object 
dtypes: float64(1), int64(1), object(4)
memory usage: 368.0+ bytes
None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   size    5 non-null      category
 1   color   5 non-null      category
 2   gender  5 non-null      category
 3   price   5 non-null      float64 
 4   weight  5 non-null      float64 
 5   bought  5 non-null      category
dtypes: category(4), float64(2)
memory usage: 740.0 bytes
None


### Zadanie 101. Przygotuj zmienne zerojedynkowe.

In [34]:
df  = pd.get_dummies(df)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        5 non-null      float64
 1   weight       5 non-null      float64
 2   size_M       5 non-null      uint8  
 3   size_XL      5 non-null      uint8  
 4   color_green  5 non-null      uint8  
 5   color_red    5 non-null      uint8  
 6   gender_male  5 non-null      uint8  
 7   bought_yes   5 non-null      uint8  
dtypes: float64(2), uint8(6)
memory usage: 238.0 bytes


### Zadanie 102. Przygotuj zmienne zerojedynkowe (n-1 poziomów.

In [33]:
df = pd.get_dummies(df,drop_first=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        5 non-null      float64
 1   weight       5 non-null      float64
 2   size_M       5 non-null      uint8  
 3   size_XL      5 non-null      uint8  
 4   color_green  5 non-null      uint8  
 5   color_red    5 non-null      uint8  
 6   gender_male  5 non-null      uint8  
 7   bought_yes   5 non-null      uint8  
dtypes: float64(2), uint8(6)
memory usage: 238.0 bytes


### Zadanie 103. Wczytaj plik typu json.

In [ ]:
df = pd.read_json('market.json')
print(df)f

### Zadanie 104. Stwórz nową zmienną w oparciu o inną.

In [ ]:
df['Kraj'] = df['Profil'].map(lambda x : x.split()[1][1:-1])
print(df)

In [35]:
x = 'Hubert Świątek'

In [40]:
df

price  weight  size_M  size_XL  color_green  color_red  gender_male  bought_yes
0  199.0   500.0       0        1            0          1            0           1
1   89.0   450.0       0        0            1          0            1           0
2   99.0   300.0       1        0            0          0            1           1
3  129.0   380.0       0        0            1          0            0           0
4   79.0   410.0       1        0            0          1            0           1

### Zadanie 105. Przygotuj listę duplikatów.

In [44]:
print(df['size_M'][df['size_M'].duplicated()].sort_values().to_list())

[0, 0, 1]


### Zadanie 106. Zmień typ zmiennej na float.

In [ ]:
df['Zmiana %']=df['Zmiana %'].map(lambda x: float(x[1:-2]))

### Zadanie 107. Wydrukuj wiersze, dla których jakąś zmienna ma najmniejszą i największą wartość.

In [ ]:
print(df.iloc[df['Zmiana %'].argmax()])
print()
print(df.iloc[df['Zmiana %'].argmin()])

### Zadanie 108. Stwórz 2 nowe obiekty typu DataFrame na podstawie istniejącego.

In [ ]:
df[df['Zmiana %']>=0].to_csv('positive.json')
df[df['Zmiana %']<0].to_csv('negative.json')

### Zadanie 109. Wczytaj dwa pliki.

In [ ]:
amazon = pd.read_csv('amazon.csv', index.col=0)
tesla = pd.read_csv('tesla.csv', index.col=0)
print(amazon)
print()
print(tesla)

### Zadanie 110. Zmień nazwy kolumn.

In [ ]:
amazon.columns = amazon.columns.map(lambda x : 'amzn_' + x.lower())
tesla.columns = tesla.columns.map(lambda x : 'tsla_' + x.lower())
print(amazon)
print()
print(tesla)

### Zadanie 111. Połącz dwie tabele.

In [49]:
result = pd.concat([amazon, tesla], axis=1)
print(result)

Index(['KKprice', 'KKweight', 'KKsize_M', 'KKsize_XL', 'KKcolor_green', 'KKcolor_red', 'KKgender_male', 'KKbought_yes'], dtype='object')

### Zadanie 112. Wyodrębnij określone kolumny.

In [ ]:
quotations = result[['amzn_open','amzn_close','tsla_open','tsla_close']]
print(quotations)

### Zadanie 113. Przygotuj kolumny wskazujące na dzienną stopę procentową.

In [ ]:
quotations['amzn_change']=(quotations['amzn_close']-quotations['amzn_open'])/quotations['amzn_open']*100
quotations['tsla_change']=(quotations['tsla_close']-quotations['tsla_open'])/quotations['tsla_open']*100
print(quotations)

### Zadanie 114. Przygotuj zmienną wskazującą na to czy w dwóch kolumnach są liczby tych samych znaków.

In [ ]:
quotations['similiarity_flag'] = 1 * (quotations['amzn_change'] * quotations['tsla_change'] > 0)

### Zadanie 115. Sprawdź jaki procent obserwacji ma daną wartość zmiennej.

In [ ]:
result = quotations['similarity_flag'].sum() / len(quotations)
print(f'{result * 100:.2f}%')

### Zadanie 116. Przygotuj obiekty typu DataFrame w oparciu o zawartości słowników.

In [58]:
products_dict = {
    'product_id': ['001', '004', '007', '010'],
    'name': ['mobile phone', 'laptop', 'mouse', 'tablet'],
    'price': [1490.0, 3400.0, 59.99, 999.00]
}

orders_day_1_dict = {
    'order_id': ['1001', '1002', '1003', '1004'],
    'product_id': ['004', '001', '001', '007'],
    'quantity': [2, 1, 1, 3]
}       

orders_day_2_dict = {
    'order_id': ['1005', '1006', '1007'],
    'product_id': ['010', '001', '007'],
    'quantity': [2, 1, 1]
}

products=pd.DataFrame(products_dict)
orders_day_1=pd.DataFrame(orders_day_1_dict)
orders_day_2=pd.DataFrame(orders_day_2_dict)
print(products)
print(orders_day_1)
print(orders_day_2)

  product_id          name    price
0        001  mobile phone  1490.00
1        004        laptop  3400.00
2        007         mouse    59.99
3        010        tablet   999.00
  order_id product_id  quantity
0     1001        004         2
1     1002        001         1
2     1003        001         1
3     1004        007         3
  order_id product_id  quantity
0     1005        010         2
1     1006        001         1
2     1007        007         1


### Zadanie 117. Przygotuj dwa obiekty typu DataFrame.

In [60]:
orders = orders_day_1.append(orders_day_2, ignore_index = True)
print(orders)

  order_id product_id  quantity
0     1001        004         2
1     1002        001         1
2     1003        001         1
3     1004        007         3
4     1005        010         2
5     1006        001         1
6     1007        007         1


C:\Users\kubeq\AppData\Local\Temp\ipykernel_15420\2443336633.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  orders = orders_day_1.append(orders_day_2, ignore_index = True)


### Zadanie 118. Połącz obiekty typu DataFrame w oparciu o tę samą nazwę kolumny.

In [61]:
result = pd.merge(orders, products, on='product_id')
print(result)

  order_id product_id  quantity          name    price
0     1001        004         2        laptop  3400.00
1     1002        001         1  mobile phone  1490.00
2     1003        001         1  mobile phone  1490.00
3     1006        001         1  mobile phone  1490.00
4     1004        007         3         mouse    59.99
5     1007        007         1         mouse    59.99
6     1005        010         2        tablet   999.00


### Zadanie 119. Utwórz kolumnę wskazującą na łączną wartość zamówienia.

In [62]:
result['total']=result['quantity']*result['price']
print(result)

  order_id product_id  quantity          name    price    total
0     1001        004         2        laptop  3400.00  6800.00
1     1002        001         1  mobile phone  1490.00  1490.00
2     1003        001         1  mobile phone  1490.00  1490.00
3     1006        001         1  mobile phone  1490.00  1490.00
4     1004        007         3         mouse    59.99   179.97
5     1007        007         1         mouse    59.99    59.99
6     1005        010         2        tablet   999.00  1998.00


### Zadanie 120. Zapisz obiekt typu DataFrame do pliku.

In [ ]:
result.to_csv('result.csv', index=False)
result.to_html('result.html')
result.to_json('result.json', indent=4)

### Zadanie 121. Połącz 2 obiekty typu DataFrame (tylko te wiersze, dla których to możliwe)

In [6]:
import pandas as pd
employees = pd.DataFrame({
    'emp_id': ['1001', '1002', '1003', '1004'],
    'first_name': ['John', 'Mark', 'Bob', 'Alice'],
    'last_name': ['Smith', 'Cook', 'Cat', 'Lee'],
    'dept_id': ['002', '001', '001', '005']
})
departments = pd.DataFrame({
    'dept_id': ['001', '002', '003', '004'],
    'dept_name': ['IT', 'Sales', 'Marketing', 'Support']
})
print(pd.merge(employees, departments, on = "dept_id", how='inner'))

  emp_id first_name last_name dept_id dept_name
0   1001       John     Smith     002     Sales
1   1002       Mark      Cook     001        IT
2   1003        Bob       Cat     001        IT


### Zadanie 122. Połącz 2 obiekty typu DataFrame (dla wierszy z pierwszej tabeli).

In [7]:
print(pd.merge(employees, departments, on = "dept_id", how='left'))

  emp_id first_name last_name dept_id dept_name
0   1001       John     Smith     002     Sales
1   1002       Mark      Cook     001        IT
2   1003        Bob       Cat     001        IT
3   1004      Alice       Lee     005       NaN


### Zadanie 123. Połącz 2 obiekty typu DataFrame (dla wierszy z drugiej tabeli).

In [8]:
print(pd.merge(employees, departments, on = "dept_id", how='left'))

  emp_id first_name last_name dept_id dept_name
0   1001       John     Smith     002     Sales
1   1002       Mark      Cook     001        IT
2   1003        Bob       Cat     001        IT
3   1004      Alice       Lee     005       NaN


### Zadanie 124. Połącz 2 obiekty typu DataFrame (dla wszystkich wierszy).

In [9]:
print(pd.merge(employees, departments, on = "dept_id", how='outer'))

  emp_id first_name last_name dept_id  dept_name
0   1001       John     Smith     002      Sales
1   1002       Mark      Cook     001         IT
2   1003        Bob       Cat     001         IT
3   1004      Alice       Lee     005        NaN
4    NaN        NaN       NaN     003  Marketing
5    NaN        NaN       NaN     004    Support


### Zadanie 125. Połącz 2 obiekty typu DataFrame (tylko te wiersze, dla których to możliwe)

In [10]:
import pandas as pd 
employees = pd.DataFrame({
    'emp_id': ['1001', '1002', '1003', '1003', '1004'],
    'first_name': ['John', 'Mark', 'Bob', 'Bob', 'Alice'],
    'last_name': ['Smith', 'Cook', 'Cat', 'Cat', 'Lee'],
    'dept_id': ['002', '001', '001', '001', '005']
})
departments = pd.DataFrame({
    'dept_id': ['001', '002', '003', '004'],
    'dept_name': ['IT', 'Sales', 'Marketing', 'Support']
})
result = pd.merge(employees, departments, how='inner', on='dept_id')
print(result)

  emp_id first_name last_name dept_id dept_name
0   1001       John     Smith     002     Sales
1   1002       Mark      Cook     001        IT
2   1003        Bob       Cat     001        IT
3   1003        Bob       Cat     001        IT


### Zadanie 126. Usuń duplikaty i połącz 2 obiekty typu DataFrame (tylko te wiersze, dla których to możliwe)

In [ ]:
employees_unique = employees.drop_duplicates()
result = pd.merge(employees_unique, departments, how='inner', on='dept_id')
print(result)

### Zadanie 127. Oblicz średnie dla podgrup.

In [14]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', 30)
pd.set_option('display.width', 1000)
tips = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv')
result = tips.groupby('day')['tip','total_bill'].mean()
result

C:\Users\kubeq\AppData\Local\Temp\ipykernel_4076\716288963.py:6: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  result = tips.groupby('day')['tip','total_bill'].mean()


tip  total_bill
day                       
Fri   2.734737   17.151579
Sat   2.993103   20.441379
Sun   3.255132   21.410000
Thur  2.771452   17.682742

### Zadanie 128. Oblicz średnie dla podgrup.

In [15]:
result = tips.groupby(['day','sex'])['tip','total_bill'].mean()
print(result)

                  tip  total_bill
day  sex                         
Fri  Female  2.781111   14.145556
     Male    2.693000   19.857000
Sat  Female  2.801786   19.680357
     Male    3.083898   20.802542
Sun  Female  3.367222   19.872222
     Male    3.220345   21.887241
Thur Female  2.575625   16.715312
     Male    2.980333   18.714667


C:\Users\kubeq\AppData\Local\Temp\ipykernel_4076\1949306885.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  result = tips.groupby(['day','sex'])['tip','total_bill'].mean()


### Zadanie 129. Wskaż wartość maksymalną dla podgrup.


In [16]:
result = tips.groupby('day')['tip','total_bill'].max()
result

C:\Users\kubeq\AppData\Local\Temp\ipykernel_4076\1616530178.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  result = tips.groupby('day')['tip','total_bill'].max()


tip  total_bill
day                    
Fri    4.73       40.17
Sat   10.00       50.81
Sun    6.50       48.17
Thur   6.70       43.11

### Zadanie 130. Oblicz średnie dla podgrup.

In [18]:
result = tips.groupby(['day','sex','smoker'])['tip','total_bill'].mean()
print(result)

                         tip  total_bill
day  sex    smoker                      
Fri  Female No      3.125000   19.365000
            Yes     2.682857   12.654286
     Male   No      2.500000   17.475000
            Yes     2.741250   20.452500
Sat  Female No      2.724615   19.003846
            Yes     2.868667   20.266667
     Male   No      3.256563   19.929063
            Yes     2.879259   21.837778
Sun  Female No      3.329286   20.824286
            Yes     3.500000   16.540000
     Male   No      3.115349   20.403256
            Yes     3.521333   26.141333
Thur Female No      2.459600   16.014400
            Yes     2.990000   19.218571
     Male   No      2.941500   18.486500
            Yes     3.058000   19.171000


C:\Users\kubeq\AppData\Local\Temp\ipykernel_4076\3294157983.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  result = tips.groupby(['day','sex','smoker'])['tip','total_bill'].mean()


### Zadanie 131. Oblicz średnie dla podgrup i pokaż wynik dla 1 podgrupy.

In [19]:
result = tips.groupby(['day','sex','smoker'])['tip','total_bill'].mean()
print(result.loc['Sun'])

                    tip  total_bill
sex    smoker                      
Female No      3.329286   20.824286
       Yes     3.500000   16.540000
Male   No      3.115349   20.403256
       Yes     3.521333   26.141333


C:\Users\kubeq\AppData\Local\Temp\ipykernel_4076\1992525591.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  result = tips.groupby(['day','sex','smoker'])['tip','total_bill'].mean()


### Zadanie 132. Oblicz średnie dla podgrup i pokaż wynik dla 1 podgrupy.

In [20]:
result = tips.groupby(['day','sex','smoker'])['tip','total_bill'].mean()
print(result.loc['Sun','Male'])

             tip  total_bill
smoker                      
No      3.115349   20.403256
Yes     3.521333   26.141333


C:\Users\kubeq\AppData\Local\Temp\ipykernel_4076\249611318.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  result = tips.groupby(['day','sex','smoker'])['tip','total_bill'].mean()
